core/src/main/java/org/apache/iceberg/MicroBatches.java

```diff
@@ -0,0 +1,266 @@
+
+      for (ManifestFile manifest : manifestFiles) {
+        manifestIndexes.add(Pair.of(manifest, currentFileIndex));
+        currentFileIndex += manifest.addedFilesCount() + manifest.existingFilesCount();
+      }
+
+      return manifestIndexes;
+    }
+
+    /**
+     * Method to skip the manifest file in which the index is smaller than startFileIndex. For example, if the
+     * index list is : (m1, 0), (m2, 3), (m3, 5), and startFileIndex is 4, then the returned manifest index list is:
+     * (m2, 3), (m3, 5).
+     *
+     * @param indexedManifests List of input manifests.
+     * @param startFileIndex Index used to skip the processed manifests.
+     * @return a sub-list of manifest file index which only contains the manifest indexes larger than the
+     * startFileIndex.
+     */
+    private List<Pair<ManifestFile, Integer>> skipManifests(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                                            int startFileIndex) {
+      if (startFileIndex == 0) {
+        return indexedManifests;
+      }
+
+      int manifestIndex = 0;
+      for (Pair<ManifestFile, Integer> manifest : indexedManifests) {
+        if (manifest.second() > startFileIndex) {
+          break;
+        }
+
+        manifestIndex++;
+      }
+
+      return indexedManifests.subList(manifestIndex - 1, indexedManifests.size());
+    }
+
+    /**
+     * Method to generate MicroBatch of this snapshot based on the indexed manifests, controlled by targetSizeInBytes.
+     *
+     * @param indexedManifests A list of indexed manifests to generate MicroBatch
+     * @param startFileIndex A startFileIndex used to skip processed files.
+     * @param targetSizeInBytes Used to control the size of MicroBatch, the processed file bytes must be smaller than
+     *                         this size.
+     * @param isStarting Used to check where all the data file should be processed, or only added files.
+     * @return A MicroBatch.
+     */
+    private MicroBatch generateMicroBatch(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                          int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      if (indexedManifests.isEmpty()) {
+        return new MicroBatch(snapshot.snapshotId(), startFileIndex, startFileIndex + 1, 0L,
+            Collections.emptyList(), true);
+      }
+
+      long currentSizeInBytes = 0L;
+      int currentFileIndex = 0;
+      boolean isLastIndex = false;
+      List<FileScanTask> tasks = Lists.newArrayList();
+
+      for (int idx = 0; idx < indexedManifests.size(); idx++) {
+        currentFileIndex = indexedManifests.get(idx).second();
+
+        try (CloseableIterable<FileScanTask> taskIterable = open(indexedManifests.get(idx).first(), isStarting);
+            CloseableIterator<FileScanTask> taskIter = taskIterable.iterator()) {
+
+          while (taskIter.hasNext()) {
```

Comment:
Because you're using try-with-resources for the Iterable, you don't really need to worry about the Iterator that was created from it. The Iterable will close all open Iterators that were created in the block. So you can write this a bit more concisely using `for`. It's fine either way, though.